FEATURE: ContentGraphAdapter for write side access

kitsunet opened the pull request:
The write side no longer uses any regular finders
for checks but only accesses projection data via
the new `ContentGraphAdapterInterface`.

Fixes: #4973

kitsunet committed:
FEATURE: ContentGraphAdapter for write side access

The write side no longer uses any regular finders
for checks but only accesses projection data via
the new `ContentGraphAdapterInterface`.

Fixes: #4973

kitsunet commented:
Few notes here for starters:

* I omitted visibility constraints everywhere as we can just assume "see everything" given that that was hardcoded before.
* I added csId + workspaceName to all the methods that need it, a tuple _might_ really make this nicer to read, but it works
* I originally thought we expose the factory and not the implementation and you can grab an implementation bound to csId + workspace _but_ there is a bunch of methods in there that do not rely on it and so it felt strange as you wouldn't have either information if you need methods that do not need it.
* it's huuuuuuge, I haven't gone over and tried to reduce it by merging similar functionality, but maybe that's an option. 
* The interface and implementation should probably move to another place, I will snatch that from bastis PR.

kitsunet commented:
Also I realized too late but why is e.g. the workspace projection in CR Core? it should be in the storage adapter implementations (so DBAL, postgres) etc. ? Same for the contentStreams...

bwaidelich commented:
> Also I realized too late but why is e.g. the workspace projection in CR Core? it should be in the storage adapter implementations

Is that related?
But for the record: Those projections work with MySQL _and_ PostgreSQL – If it _was_ part of the storage adapter we would need to copy all of the implementations to both adapters.
But I agree that it should be in separate packages

kitsunet commented:
More note.... I guess the adapter needs a bunch more dependencies to actually be able to provide Node(s) results, we'll see how that works out, they are obviously all available when building it, but I would rather want to avoid adding them...

bwaidelich commented on Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Repository/ContentGraphAdapter.php:
```diff
@@ -0,0 +1,123 @@
+use Neos\ContentRepository\Core\SharedModel\Workspace\WorkspaceName;
+
+/**
+ *
```
This needs some thorough documentation and an `@internal` annotation

kitsunet replied:
Yep, documentation is missing _everywhere_ just wanted to get this out here for now. Important though, so thanks!

kitsunet commented:
> > Also I realized too late but why is e.g. the workspace projection in CR Core? it should be in the storage adapter implementations
> 
> Is that related? 

I tried to get rid of ALL access to finders within the handlers, so now I also took the worksapce and contentstream related methods into the adapter, which doesn't work because the DBAL adapter implementation then doesn't know about those :D

bwaidelich commented on Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Repository/ContentGraphAdapter.php:
```diff
@@ -0,0 +1,123 @@
+        // TODO: Implement rootNodeAggregateWithTypeExists() method.
+    }
+
+    public function findParentNodeAggregates(ContentStreamId $contentStreamId, WorkspaceName $workspaceName, NodeAggregateId $childNodeAggregateId): iterable
```
Not directly related but we should really introduce a `NodeAggregates` DTO

kitsunet replied:
Yes, can do maybe

bwaidelich commented on Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Repository/ContentGraphAdapterFactory.php:
```diff
@@ -0,0 +1,27 @@
+use Neos\ContentRepository\Core\SharedModel\ContentRepository\ContentRepositoryId;
+
+/**
+ * This factory
```
...?  :)

kitsunet replied:
...needs more documentation 😆 see above

bwaidelich commented on Neos.ContentRepository.Core/Classes/Feature/Common/ConstraintChecks.php:
```diff
@@ -387,11 +387,12 @@ protected function areNodeTypeConstraintsImposedByGrandparentValid(
     ): NodeAggregate {
-        $nodeAggregate = $contentRepository->getContentGraph()->findNodeAggregateById(
+        $workspace = $this->getContentGraphAdapter()->findWorkspaceByCurrentContentStreamId($contentStreamId);
+        $nodeAggregate = $this->getContentGraphAdapter()->findNodeAggregateById(
```
IMO one of the goals of the adapter should be that we _don't_ have to fetch the workspace in order to retrieve node aggregate details.
The `ContentStreamIdOverride` should happen _in_ the adapter (or rather: we should probably fetch a different instance if the override is active)

kitsunet replied:
Mmm, that is also an option obviously, but then you have zero outside control, that is, you could _in theory_ get nodes with one workspace and then the other a bit later? Lets discuss.

kitsunet committed:
Make ContentGraphAdapterFactory configurable

bwaidelich reviewed (commented):
Thanks so much for your efforts!
I just did a first quick pass as preparation for our weekly.

Apart from the nitpicky comments, I think the adapter should always act on a workspace/contentstream

kitsunet committed:
Introduce ContentGraphAdapterProvider

kitsunet committed:
Introduce ContentGraphAdapterProviderInterface

This also replaces the ContentStreamIdOverride

kitsunet commented on Neos.ContentRepository.Core/Classes/Feature/ContentGraphAdapterInterface.php:
```diff
@@ -0,0 +1,172 @@
+        ContentStreamId $contentStreamId
+    ): ?Workspace;
+
+    public function withWorkspaceName(WorkspaceName $workspaceName): ContentGraphAdapterInterface;
```
nope, we have the adapter for this now, this was an inbetween thing.

kitsunet commented on Neos.ContentRepository.Core/Classes/Feature/ContentGraphAdapterInterface.php:
```diff
@@ -0,0 +1,172 @@
+
+    public function withWorkspaceName(WorkspaceName $workspaceName): ContentGraphAdapterInterface;
+
+    public function getWorkspaceName(): WorkspaceName;
```
These getters for the WorkspaceName and ContentStreamId are super useful if you follow the trail throughout the nested functions in traits and such. I think it makes sense.

kitsunet commented on Neos.ContentRepository.Core/Classes/Feature/ContentGraphAdapterInterface.php:
```diff
@@ -0,0 +1,172 @@
+     * WORKSPACES
+     */
+
+    public function findWorkspaceByCurrentContentStreamId(
```
This also is unused now and will be removed in the next update

kitsunet commented on Neos.ContentRepository.Core/Classes/Feature/ContentGraphAdapterInterface.php:
```diff
@@ -0,0 +1,172 @@
+     * CONTENT STREAMS
+     */
+
+    public function hasContentStream(): bool;
```
as outlined by the ContentGraphAdapterProviderInterface I think that one should just throw and not even offer us an adapter if there is no content stream?

kitsunet committed:
ContentGraphAdapter working implementation

Adds the working DBAL implementation as well as some fixes.
Also new Factory/Builder concept

kitsunet committed:
Merge branch '9.0' into feature/content-graph-adapter-4973

kitsunet committed:
Linting and Naming ContentGraphAdapter

kitsunet committed:
More code cleanup

kitsunet commented:
> Now come the hard(er) parts in a way. The nitty gritty. Some of the _missing_ methods (see linter errors) were removed by me because phpstorm didn't detect usage. The interesting question is which of those use cases do we consider valid "read" use cases that should use the content graph and which should use the new ContentGraphAdapter. I think the two controllers are probably _real_ use cases and should be re-instated to work. the ContentCacheFlusher for example I would rather solve with the ContentGraphAdapter as well as the SiteServiceInternals...

All resolved for now, but I will probably poke a few other corners. Also now need to deduplicate the queries.

kitsunet committed:
Code renaming and refinement

kitsunet committed:
linter fix

kitsunet committed:
Use ContentGraphAdapter in SiteServiceInternals

kitsunet committed:
Query deduplication part 1

kitsunet committed:
Query deduplication part 2

kitsunet committed:
More cleanup

kitsunet committed:
Fix that phpstan doesn't understand condition

kitsunet commented on Neos.ContentGraph.DoctrineDbalAdapter/src/ContentGraphAdapterFactoryBuilder.php:
```diff
@@ -0,0 +1,23 @@
+ * Builder to combine injected dependencies and ProjectionFActoryDependencies into a ContentGraphAdapterFactory
+ * @internal
+ */
+class ContentGraphAdapterFactoryBuilder implements ContentGraphAdapterFactoryBuilderInterface
```
I wouldn't mind getting around having this, and am happy for suggestions, I don't really see it though. An ugly option would be to to have something along the lines of:

`ContentGraphAdapterFactory::injectProjectionDependencies(ProjectionFactoryDependencies $projectionFactoryDependencies)`

and the `ContentGraphAdapterFactory::__construct` just gets the adapter dependencies injected (eg. DBALClient), but then all methods in the ContentGraphAdapterFactory need to throw a "TooEarly" Exception if the projection dependencies were not injected yet.

bwaidelich replied:
AFAIS we only need it to pass in `ContentRepositoryId`, `NodeTypeManager` and `PropertyConverter` – Maybe we can make these explicit dependencies somehow

bwaidelich replied:
..or even make the `NodeFactory` an explicit dependency (via some `NodeFactoryInterface`) but I don't know... Let's discuss

kitsunet commented on Neos.ContentRepository.TestSuite/Classes/Behavior/Features/Bootstrap/GenericCommandExecutionAndEventPublication.php:
```diff
@@ -224,6 +224,8 @@ public function eventNumberIs(int $eventNumber, string $eventType, TableNode $pa
                 $actual = DimensionSpacePointSet::fromArray($actualValue);
                 Assert::assertTrue($expected->equals($actual), 'Actual Dimension Space Point set "' . json_encode($actualValue) . '" does not match expected Dimension Space Point set "' . $assertionTableRow['Expected'] . '"');
             } else {
+//                die();
```
weird, I am not even sure why I woudl want to do something here^^ Removing

bwaidelich replied:
?

bwaidelich commented on Neos.ContentRepository.TestSuite/Classes/Behavior/Features/Bootstrap/GenericCommandExecutionAndEventPublication.php:
```diff
@@ -224,6 +224,8 @@ public function eventNumberIs(int $eventNumber, string $eventType, TableNode $pa
                 $actual = DimensionSpacePointSet::fromArray($actualValue);
                 Assert::assertTrue($expected->equals($actual), 'Actual Dimension Space Point set "' . json_encode($actualValue) . '" does not match expected Dimension Space Point set "' . $assertionTableRow['Expected'] . '"');
             } else {
+//                die();
+//                exit;
```
Ah you mean, you want to remove those changes again? I agree :)
```suggestion
```

kitsunet committed:
Cleanup

kitsunet committed:
Make factory signature clearer

kitsunet committed:
Linter fixes

kitsunet committed:
Merge branch '9.0' into feature/content-graph-adapter-4973

kitsunet committed:
Add ContentGraphTableNames

I think the idea is good, just where we create it
and how we pass it along needs refinement, given
that this PR will not stay as it is anyways, just
putting it in to keep the idea intact.

kitsunet commented:
In our weekly 2024-04-19 we talked about how this adapter is maybe not worth the trouble and if we can solve the ugly contentStreamIdOverride differently without this. And it is indeed difficult to make assumptions for after introduction of node identity, but it is very clear to me that at a higher level commands and API will use workspaceName the queries (at least for the DBAL Adapter) will need a contentStreamId to work, therefore this translation will have to happen somewhere, IMHO the changes for NodeIdentity will just push this deeper into the read side making it even more annoying (e.g. due to`ContentSubgraphWithRuntimeCaches`) to override for the few caess that we have. And those are valid cases as we apply commands with $some workspaceName to a contentStream that has (not yet got) this workspaceName.

This together with the (from the POV of the write side) convoluted API to read stuff lets me think that this change could have merit even IF we cannot at this point avoid using read models (IMHO this could be a later improvement though) and much easier to just change the write side interface for this.

Convoluted because we apply filters, visibility and a lot of other stuff.

I am still not in love with the triple factory thing, but the separation of APIs seems very sensible still.

mhsdesign commented on Neos.ContentRepositoryRegistry.DoctrineDbalClient/Configuration/Settings.yaml:
```diff
@@ -7,3 +7,5 @@ Neos:
           # catchUpHooks for content cache flushing
           'Neos.ContentRepository:ContentGraph':
             factoryObjectName: Neos\ContentGraph\DoctrineDbalAdapter\DoctrineDbalContentGraphProjectionFactory
+        contentGraphAdapterFactory:
+          factoryObjectName: Neos\ContentGraph\DoctrineDbalAdapter\ContentGraphAdapterFactoryBuilder
```
optional followup would be to simplify this config

mhsdesign commented on Neos.ContentRepositoryRegistry/Classes/ContentRepositoryRegistry.php:
```diff
@@ -281,4 +285,16 @@ private function buildClock(ContentRepositoryId $contentRepositoryIdentifier, ar
+        if (!$contentGraphAdapterFactoryBuilder instanceof ContentGraphAdapterFactoryBuilderInterface) {
+            throw InvalidConfigurationException::fromMessage('contentGraphAdapterFactory.factoryObjectName for content repository "%s" is not an instance of %s but %s.', $contentRepositoryIdentifier->value, ContentGraphAdapterFactoryBuilderInterface::class, get_debug_type($contentGraphAdapterFactoryBuilder));
+        }
+        // TODO: Do we want to add options here? Would then have to be a setter I guess....
```
todo over here ;)

mhsdesign commented on Neos.ContentGraph.DoctrineDbalAdapter/src/ContentGraphAdapter.php:
We should refine this naming

mhsdesign commented on Neos.ContentGraph.DoctrineDbalAdapter/src/ContentGraphAdapter.php:
```diff
@@ -0,0 +1,507 @@
+        return $this->contentStreamId;
+    }
+
+    public function getWorkspace(): Workspace
+    {
+        $query = $this->dbalConnection->prepare('SELECT * FROM cr_default_p_workspace WHERE workspacename LIKE :workspaceName');
```
illegal select, and the `Workspace` tdo is too big

mhsdesign commented on Neos.ContentGraph.DoctrineDbalAdapter/src/ContentGraphAdapter.php:
```diff
@@ -0,0 +1,507 @@
+            return $this->contentStreamId;
+        }
+
+        // TODO: This table is not allowed here...
+        $query = $this->dbalConnection->prepare('SELECT currentcontentstreamid FROM cr_default_p_workspace WHERE workspacename LIKE :workspaceName');
```
todo illegal

bwaidelich commented:
This is to be replaced by #5028 and I suggest to close the PR (we can still keep the branch for reference if needed)
